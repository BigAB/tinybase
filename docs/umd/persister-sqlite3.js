var a,t;a=this,t=function(a){"use strict";const t=a=>typeof a,e="tinybase",s=",",n=t(""),i=(a,t)=>a.repeat(t),c=(a,t="")=>a.join(t),o=(a,t)=>a.map(t),r=a=>a.length,l=a=>0==r(a),y=(a,t)=>a.filter(t),w=(a,t,e)=>a.slice(t,e),u=(a,...t)=>a.push(...t),E=Promise,d=(a,t)=>a instanceof t,f=a=>null==a,p=(a,t,e)=>f(a)?e?.():t(a),T=a=>t(a)==n,m=async a=>E.all(a),v=(a,t)=>a?.has(t)??!1,A=a=>[...a?.values()??[]],L=(a,t)=>a?.delete(t),h=Object,N=h.keys,R=h.freeze,O=(a=[])=>h.fromEntries(a),S=(...a)=>h.assign({},...a),C=(a,t)=>o(h.entries(a),(([a,e])=>t(e,a))),g=a=>h.values(a),b=a=>r(N(a)),D=a=>(a=>d(a,h)&&a.constructor==h)(a)&&0==b(a),I=a=>new Map(a),F=a=>[...a?.keys()??[]],M=(a,t)=>a?.get(t),$=(a,t)=>o([...a?.entries()??[]],(([a,e])=>t(e,a))),P=(a,t,e)=>f(e)?(L(a,t),a):a?.set(t,e),_=(a,t,e)=>(v(a,t)||P(a,t,e()),M(a,t)),B=(a,t,e,s=P)=>(C(t,((t,s)=>e(a,s,t))),((a,t)=>{((a,t)=>{a?.forEach(t)})(a,((a,e)=>t(e)))})(a,(e=>((a,t)=>!f(((a,t)=>p(a,(a=>a[t])))(a,t)))(t,e)?0:s(a,e))),a),j=a=>new Set(Array.isArray(a)||f(a)?a:[a]),x=(a,t)=>a?.add(t),H="_",W="_id",k=a=>`"${a.replace(/"/g,'""')}"`,q="FROM pragma_table_",J="WHERE",U=(a,t,e)=>{const n=I();return[async()=>B(n,O(await m(o(await a("SELECT name "+q+"list WHERE schema='main'AND type='table'AND name IN("+z(t)+")",t),(async({name:t})=>[t,O(o(await a("SELECT name,type "+q+"info(?)",[t]),(({name:a,type:t})=>[a,t])))])))),((a,t,e)=>P(n,t,B(_(n,t,I),e,((a,t,e)=>{e!=M(a,t)&&P(a,t,e)}),((a,t)=>P(a,t))))),((a,t)=>P(n,t))),async(t,e)=>((a,t)=>!f(M(M(n,a),t)))(t,e)?O(y(o(await a("SELECT*FROM"+k(t)),(a=>{return[a[e],(t={...a},s=e,delete t[s],t)];var t,s})),(([a,t])=>!f(a)&&!D(t)))):{},async(t,e,i,r,w,E=!1)=>{const d=j();C(i??{},(a=>o(N(a??{}),(a=>x(d,a)))));const p=A(d);if(!E&&w&&l(p)&&v(n,t))return await a("DROP TABLE"+k(t)),void P(n,t);if(l(p)||v(n,t)){const s=M(n,t),i=j(F(s));await m([...o(p,(async e=>{L(i,e)||(await a(`ALTER TABLE${k(t)}ADD${k(e)}`),P(s,e,""))})),...!E&&r?o(A(i),(async n=>{n!=e&&(await a(`ALTER TABLE${k(t)}DROP${k(n)}`),P(s,n))})):[]])}else await a(`CREATE TABLE${k(t)}(${k(e)} PRIMARY KEY ON CONFLICT REPLACE${c(o(p,(a=>s+k(a))))});`),P(n,t,I([[e,""],...o(p,(a=>[a,""]))]));if(E)f(i)?await a("DELETE FROM"+k(t)+"WHERE 1"):await m(C(i,(async(s,n)=>{f(s)?await a("DELETE FROM"+k(t)+J+k(e)+"=?",[n]):l(p)||await Y(a,t,e,N(s),[n,...g(s)])})));else if(l(p))v(n,t)&&await a("DELETE FROM"+k(t)+"WHERE 1");else{const s=y(F(M(n,t)),(a=>a!=e)),c=[],r=[];C(i??{},((a,t)=>{u(c,t,...o(s,(t=>a?.[t]))),u(r,t)})),await Y(a,t,e,s,c),await a("DELETE FROM"+k(t)+J+k(e)+"NOT IN("+z(r)+")",r)}},async t=>{let s;await a("BEGIN");try{s=await t()}catch(a){e?.(a)}return await a("END"),s}]},Y=async(a,t,e,n,l)=>await a("INSERT INTO"+k(t)+"("+k(e)+c(o(n,(a=>s+k(a))))+")VALUES"+i(`,(?${i(",?",r(n))})`,r(l)/(r(n)+1)).substring(1)+"ON CONFLICT("+k(e)+")DO UPDATE SET"+c(o(n,(a=>k(a)+"=excluded."+k(a))),s),l),z=a=>c(o(a,(()=>"?")),s),G=JSON.parse,K=I(),V=I(),Q=(a,t,e,s,n,i,c=[])=>{let o,r,l,y=0,w=0;_(K,c,(()=>0)),_(V,c,(()=>[]));const E=async a=>(2!=y&&(y=1,await d.schedule((async()=>{await a(),y=0}))),d),d={load:async(e,s)=>await E((async()=>{try{a.setContent(await t())}catch{a.setContent([e,s])}})),startAutoLoad:async(e={},n={})=>(d.stopAutoLoad(),await d.load(e,n),w=1,l=s((async(e,s)=>{if(s){const t=s();await E((async()=>a.setTransactionChanges(t)))}else await E((async()=>{try{a.setContent(e?.()??await t())}catch(a){i?.(a)}}))})),d),stopAutoLoad:()=>(w&&(n(l),l=void 0,w=0),d),save:async t=>(1!=y&&(y=2,await d.schedule((async()=>{try{await e(a.getContent,t)}catch(a){i?.(a)}y=0}))),d),startAutoSave:async()=>(await d.stopAutoSave().save(),o=a.addDidFinishTransactionListener(((a,t)=>{const[e,s]=t();D(e)&&D(s)||d.save((()=>[e,s]))})),d),stopAutoSave:()=>(p(o,a.delListener),d),schedule:async(...a)=>(u(M(V,c),...a),await(async()=>{if(!M(K,c)){for(P(K,c,1);!f((a=M(V,c),r=a.shift()));)try{await r()}catch(a){i?.(a)}P(K,c,0)}var a})(),d),getStore:()=>a,destroy:()=>d.stopAutoLoad().stopAutoSave(),getStats:()=>({})};return R(d)},X="store",Z=(a,t,e,s,n,[i],c,o)=>{const[r,l,y,w]=U(t,c,n);return Q(a,(async()=>await w((async()=>(await r(),G((await l(i,W))[H]?.[X]??"null"))))),(async a=>await w((async()=>{var t;await r(),await y(i,W,{[H]:{[X]:(t=a()??null,JSON.stringify(t,((a,t)=>d(t,Map)?h.fromEntries([...t]):t)))}},!0,!0)}))),e,s,n,o)},aa=(a,t,e,s,n,[i,c,[o,r,l]],w,u)=>{const[E,d,p,T]=U(t,w,n),v=async(a,t)=>await m($(c,(async([e,s,n,i],c)=>{const o=a[c];t&&void 0===o||await p(e,s,o,n,i,t)}))),A=async(a,t)=>r?await p(l,W,{[H]:a},!0,!0,t):null;return Q(a,(async()=>await T((async()=>{await E();const a=await(async()=>O(y(await m($(i,(async([a,t],e)=>[a,await d(e,t)]))),(a=>!D(a[1])))))(),t=await(async()=>o?(await d(l,W))[H]:{})();return D(a)&&f(t)?void 0:[a,t]}))),(async(a,t)=>await T((async()=>{if(await E(),f(t)){const[t,e]=a();await v(t),await A(e)}else{const[a,e]=t();await v(a,!0),await A(e,!0)}}))),e,s,n,u)},ta="json",ea="autoLoadIntervalSeconds",sa="rowIdColumnName",na="tableId",ia="tableName",ca={mode:ta,[ea]:1},oa={load:0,save:0,[ia]:e+"_values"},ra=(a,t,e,s)=>{const n=I();return C(a,((a,i)=>{const c=w(g(S(t,T(a)?{[e]:a}:a)),0,b(t));f(c[0])||s(i,c[0])||P(n,i,c)})),n},la="pragma ",ya="data_version",wa="schema_version",ua=(a,t,s,n,i,c,o,r)=>{let l,y;const[u,E,d,f]=(a=>{const t=(a=>S(ca,T(a)?{storeTableName:a}:a??{}))(a),s=t[ea];if(t.mode==ta){const{storeTableName:a=e}=t;return[1,s,[a],j(a)]}const{tables:{load:n={},save:i={}}={},values:c={}}=t,o=w(g(S(oa,c)),0,b(oa)),r=o[2],l=j(r);return[0,s,[ra(n,{[na]:null,[sa]:W},na,(a=>x(l,a)&&a==r)),ra(i,{[ia]:null,[sa]:W,deleteEmptyColumns:0,deleteEmptyTable:0},ia,((a,t)=>x(l,t)&&t==r)),o],l]})(t);return(u?Z:aa)(a,c?async(a,t)=>(c(a,t),await s(a,t)):s,(a=>[setInterval((async()=>{try{const t=(await s(la+ya))[0][ya],e=(await s(la+wa))[0][wa];t==(l??=t)&&e==(y??=e)||(a(),l=t,y=e)}catch{}}),1e3*E),n((t=>f.has(t)?a():0))]),(([a,t])=>{clearInterval(a),l=y=null,i(t)}),o,d,A(f),r)},Ea="change";a.createSqlite3Persister=(a,t,e,s,n)=>ua(a,e,(async(a,e=[])=>{return await(s=(s,n)=>t.all(a,e,((a,t)=>a?n(a):s(t.map((a=>({...a})))))),new E(s));var s}),(a=>{const e=(t,e,s)=>a(s);return t.on(Ea,e),e}),(a=>t.off(Ea,a)),s,n,t)},"object"==typeof exports&&"undefined"!=typeof module?t(exports):"function"==typeof define&&define.amd?define(["exports"],t):t((a="undefined"!=typeof globalThis?globalThis:a||self).TinyBasePersisterSqlite3={});
