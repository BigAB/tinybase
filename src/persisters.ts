import {DEBUG, ifNotUndefined, isUndefined} from './common/other';
import {GetTransactionChanges, Store, Tables, Values} from './types/store.d';
import {
  Persister,
  PersisterListener,
  PersisterStats,
} from './types/persisters.d';
import {arrayPush, arrayShift} from './common/array';
import {Id} from './types/common.d';
import {objFreeze} from './common/obj';

type Action = () => Promise<any>;

export const createCustomPersister = <ListeningHandle>(
  store: Store,
  getPersisted: () => Promise<[Tables, Values] | undefined>,
  setPersisted: (
    getContent: () => [Tables, Values],
    getTransactionChanges?: GetTransactionChanges,
  ) => Promise<void>,
  addPersisterListener: (listener: PersisterListener) => ListeningHandle,
  delPersisterListener: (listeningHandle: ListeningHandle) => void,
): Persister => {
  let listenerId: Id | undefined;
  let loadSave = 0;
  let loads = 0;
  let saves = 0;
  let running = 0;
  let listening = 0;
  let action;
  let listeningHandle: ListeningHandle | undefined;

  const scheduledActions: Action[] = [];

  const run = async (): Promise<void> => {
    /*! istanbul ignore else */
    if (!running) {
      running = 1;
      while (!isUndefined((action = arrayShift(scheduledActions)))) {
        try {
          await action();
        } catch {}
      }
      running = 0;
    }
  };

  const loadLock = async (actions: Action) => {
    /*! istanbul ignore else */
    if (loadSave != 2) {
      loadSave = 1;
      if (DEBUG) {
        loads++;
      }
      await actions();
      loadSave = 0;
    }
  };

  const persister: any = {
    load: async (
      initialTables: Tables = {},
      initialValues: Values = {},
    ): Promise<Persister> => {
      await loadLock(async () => {
        try {
          store.setContent((await getPersisted()) as [Tables, Values]);
        } catch {
          store.setContent([initialTables, initialValues]);
        }
      });
      return persister;
    },

    startAutoLoad: async (
      initialTables: Tables = {},
      initialValues: Values = {},
    ): Promise<Persister> => {
      persister.stopAutoLoad();
      await persister.load(initialTables, initialValues);
      listening = 1;
      listeningHandle = addPersisterListener(
        async (getContent, getTransactionChanges) => {
          await loadLock(async () => {
            if (getTransactionChanges) {
              store.setTransactionChanges(getTransactionChanges());
            } else {
              try {
                store.setContent(
                  getContent?.() ??
                    ((await getPersisted()) as [Tables, Values]),
                );
              } catch {}
            }
          });
        },
      );
      return persister;
    },

    stopAutoLoad: (): Persister => {
      if (listening) {
        delPersisterListener(listeningHandle as ListeningHandle);
        listeningHandle = undefined;
        listening = 0;
      }
      return persister;
    },

    save: async (
      getTransactionChanges?: GetTransactionChanges,
    ): Promise<Persister> => {
      await persister.schedule(async () => {
        /*! istanbul ignore else */
        if (loadSave != 1) {
          loadSave = 2;
          if (DEBUG) {
            saves++;
          }
          try {
            await setPersisted(store.getContent, getTransactionChanges);
          } catch {}
          loadSave = 0;
        }
      });
      return persister;
    },

    startAutoSave: async (): Promise<Persister> => {
      await persister.stopAutoSave().save();
      listenerId = store.addDidFinishTransactionListener(
        (_store, getTransactionChanges) => {
          const transactionChanges = getTransactionChanges();
          (persister.save as any)(() => transactionChanges);
        },
      );
      return persister;
    },

    stopAutoSave: (): Persister => {
      ifNotUndefined(listenerId, store.delListener);
      return persister;
    },

    getStore: (): Store => store,

    destroy: (): Persister => persister.stopAutoLoad().stopAutoSave(),

    getStats: (): PersisterStats => (DEBUG ? {loads, saves} : {}),

    schedule: async (...actions: Action[]): Promise<void> => {
      arrayPush(scheduledActions, ...actions);
      await run();
    },
  };

  return objFreeze(persister as Persister);
};
