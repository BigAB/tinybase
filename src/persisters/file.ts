import {Callback, Json} from '../types/common.d';
import {
  Persister,
  createFilePersister as createFilePersisterDecl,
} from '../types/persisters.d';
import {promises, watch} from 'fs';
import {FSWatcher} from 'fs';
import {Store} from '../types/store.d';
import {UTF8} from '../common/strings';
import {createCustomPersister} from './common';

export const createFilePersister = ((
  store: Store,
  filePath: string,
): Persister => {
  const getPersisted = async (): Promise<string | null | undefined> => {
    try {
      return await promises.readFile(filePath, UTF8);
    } catch {}
  };

  const setPersisted = async (json: Json): Promise<void> => {
    try {
      await promises.writeFile(filePath, json, UTF8);
    } catch {}
  };

  const startListeningToPersisted = (didChange: Callback): FSWatcher =>
    watch(filePath, didChange);

  const stopListeningToPersisted = (watcher: FSWatcher): void =>
    watcher?.close();

  return createCustomPersister(
    store,
    getPersisted,
    setPersisted,
    startListeningToPersisted,
    stopListeningToPersisted,
  );
}) as typeof createFilePersisterDecl;
