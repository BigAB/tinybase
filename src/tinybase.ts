export * from './metrics';
export * from './store';
