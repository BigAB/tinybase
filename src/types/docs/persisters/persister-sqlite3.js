/**
 * The persister-sqlite3 module of the TinyBase project lets you save and load
 * Store data to and from a local SQLite database (in an appropriate
 * environment).
 *
 * @see Persisting Data guide
 * @packageDocumentation
 * @module persister-sqlite3
 */
/// persister-sqlite3
/**
 * The createSqlite3Persister function creates a Persister object that can
 * persist the Store to a local SQLite database (in an appropriate environment).
 *
 * As well as providing a reference to the Store to persist, you must provide a
 * `db` parameter which identifies the database instance.
 *
 * @param store The Store to persist.
 * @param db The database instance that was returned from `new
 * sqlite3.Database(...)`.
 * @returns A reference to the new Persister object.
 * @example
 * This example creates a Persister object and persists the Store to a local
 * SQLite database.
 *
 * ```js yolo
 * const db = new sqlite3.Database('/db.sqlite3');
 *
 * const store = createStore().setTables({pets: {fido: {species: 'dog'}}});
 * const persister = createSqlite3Persister(store, db);
 *
 * await persister.save();
 * // Store JSON will be saved to the database.
 *
 * await persister.load();
 * // Store JSON will be loaded from the database.
 *
 * persister.destroy();
 * ```
 * @category Creation
 */
/// createSqlite3Persister
