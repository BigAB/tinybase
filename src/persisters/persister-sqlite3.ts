import {DatabasePersisterConfig, Persister} from '../types/persisters';
import {UpdateListener, createSqlitePersister} from './sqlite/create';
import {Database} from 'sqlite3';
import {IdObj} from '../common/obj';
import {Store} from '../types/store';
import {createSqlite3Persister as createSqlite3PersisterDecl} from '../types/persisters/persister-sqlite3';
import {promiseNew} from '../common/other';

const CHANGE = 'change';

type Observer = (_: any, _2: any, tableName: string) => void;

export const createSqlite3Persister = ((
  store: Store,
  db: Database,
  configOrStoreTableName?: DatabasePersisterConfig | string,
): Persister =>
  createSqlitePersister(
    store,
    configOrStoreTableName,
    (sql: string, args: any[] = []): Promise<IdObj<any>[]> =>
      promiseNew((resolve, reject) =>
        db.all(sql, args, (error, rows: IdObj<any>[]) =>
          error
            ? reject(error)
            : resolve(rows.map((row: IdObj<any>) => ({...row}))),
        ),
      ),
    (listener: UpdateListener): Observer => {
      const observer = (_: any, _2: any, tableName: string) =>
        listener(tableName);
      db.on(CHANGE, observer);
      return observer;
    },
    (observer: Observer): any => db.off(CHANGE, observer),
  )) as typeof createSqlite3PersisterDecl;
