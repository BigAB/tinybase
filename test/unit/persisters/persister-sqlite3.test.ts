/* eslint-disable max-len */
import {Persister, Store, createStore} from 'tinybase/debug';
import {mkdirSync, unlinkSync} from 'fs';
import sqlite3, {Database} from 'sqlite3';
import {createSqlite3Persister} from 'tinybase/debug/persisters/persister-sqlite3';
import {dirname} from 'path';

type Dump = [name: string, sql: string, rows: {[column: string]: any}[]][];

const TEST_FILE = 'tmp/test.sqlite3';

const escapeId = (str: string) => `"${str.replace(/"/g, '""')}"`;

const exec = (sql: string, args: any[] = []): Promise<void> =>
  new Promise((resolve) => db.run(sql, args, () => resolve()));

const get = (sql: string, args: any[] = []): Promise<{[id: string]: any}[]> =>
  new Promise((resolve) =>
    db.all(sql, args, (_, rows: {[id: string]: any}[]) => resolve(rows)),
  );

const getDatabase = async (): Promise<Dump> =>
  await Promise.all(
    (
      await get(`SELECT sql, name FROM sqlite_schema WHERE type = 'table'`)
    ).map(async ({sql, name}: any) => [
      name,
      sql,
      await get('SELECT * FROM ' + escapeId(name)),
    ]),
  );

let db: Database;
let store: Store;
let persister: Persister;

const setDatabase = async (dump: Dump) =>
  dump.forEach(async ([name, sql, rows]) => {
    await exec(sql);
    await Promise.all(
      rows.map(
        async (row) =>
          await exec(
            'INSERT INTO ' +
              name +
              '(' +
              Object.keys(row).join(',') +
              ') VALUES (' +
              Object.keys(row)
                .map(() => '?')
                .join(',') +
              ')',
            Object.values(row),
          ),
      ),
    );
  });

beforeEach(async () => {
  try {
    mkdirSync(dirname(TEST_FILE), {recursive: true});
    unlinkSync(TEST_FILE);
  } catch {}
  db = new sqlite3.Database(TEST_FILE);
  store = createStore();
});

afterEach(() => {
  db.close();
});

describe('Serialized', () => {
  beforeEach(async () => {
    persister = createSqlite3Persister(store, db);
  });

  describe('Custom table name', () => {
    test('as string', async () => {
      const persister = createSqlite3Persister(store, db, 'test');
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'test',
          'CREATE TABLE "test"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{},{}]'}],
        ],
      ]);
    });

    test('with spaces', async () => {
      const persister = createSqlite3Persister(store, db, 'test table');
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'test table',
          'CREATE TABLE "test table"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{},{}]'}],
        ],
      ]);
    });

    test('with quote', async () => {
      const persister = createSqlite3Persister(store, db, 'test "table"');
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'test "table"',
          'CREATE TABLE "test ""table"""("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{},{}]'}],
        ],
      ]);
    });

    test('as config', async () => {
      const persister = createSqlite3Persister(store, db, {
        serialized: true,
        storeTable: 'test',
      });
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'test',
          'CREATE TABLE "test"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{},{}]'}],
        ],
      ]);
    });
  });

  describe('Save to empty database', () => {
    test('nothing', async () => {
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{},{}]'}],
        ],
      ]);
    });

    test('tables', async () => {
      store.setTables({t1: {r1: {c1: 1}}});
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{"t1":{"r1":{"c1":1}}},{}]'}],
        ],
      ]);
    });

    test('values', async () => {
      store.setValues({v1: 1});
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{},{"v1":1}]'}],
        ],
      ]);
    });

    test('both', async () => {
      store.setTables({t1: {r1: {c1: 1}}}).setValues({v1: 1});
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{"t1":{"r1":{"c1":1}}},{"v1":1}]'}],
        ],
      ]);
    });
  });

  describe('Load from database', () => {
    test('nothing', async () => {
      await persister.load();
      expect(store.getContent()).toEqual([{}, {}]);
    });

    test('defaulted', async () => {
      await persister.load({t1: {r1: {c1: 1}}}, {v1: 1});
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {v1: 1}]);
    });

    test('broken', async () => {
      setDatabase([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{"t1": 1}]'}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{}, {}]);
    });

    test('broken, can default', async () => {
      setDatabase([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{"t1": }]'}],
        ],
      ]);
      await persister.load({t1: {r1: {c1: 1}}}, {v1: 1});
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {v1: 1}]);
    });

    test('tables', async () => {
      setDatabase([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{"t1": {"r1": {"c1": 1}}}, {}]'}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {}]);
    });

    test('values', async () => {
      setDatabase([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{}, {"v1": 1}]'}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{}, {v1: 1}]);
    });

    test('both', async () => {
      setDatabase([
        [
          'tinybase',
          'CREATE TABLE "tinybase"("_id" PRIMARY KEY ON CONFLICT REPLACE, "store")',
          [{_id: '_', store: '[{"t1": {"r1": {"c1": 1}}}, {"v1": 1}]'}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {v1: 1}]);
    });
  });
});

describe('Non-serialized', () => {
  beforeEach(async () => {
    persister = createSqlite3Persister(store, db, {serialized: false});
  });

  describe('Custom row id column', () => {
    test('word', async () => {
      const persister = createSqlite3Persister(store, db, {
        serialized: false,
        rowIdColumn: 'test',
      });
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("test" PRIMARY KEY ON CONFLICT REPLACE)',
          [{test: '_'}],
        ],
      ]);
    });

    test('with spaces', async () => {
      const persister = createSqlite3Persister(store, db, {
        serialized: false,
        rowIdColumn: 'test table',
      });
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("test table" PRIMARY KEY ON CONFLICT REPLACE)',
          [{'test table': '_'}],
        ],
      ]);
    });

    test('with quote', async () => {
      const persister = createSqlite3Persister(store, db, {
        serialized: false,
        rowIdColumn: 'test "table"',
      });
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("test ""table""" PRIMARY KEY ON CONFLICT REPLACE)',
          [{'test "table"': '_'}],
        ],
      ]);
    });
  });

  describe('Custom values table name', () => {
    test('word', async () => {
      const persister = createSqlite3Persister(store, db, {
        serialized: false,
        valuesTable: 'test',
      });
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'test',
          'CREATE TABLE "test"("_id" PRIMARY KEY ON CONFLICT REPLACE)',
          [{_id: '_'}],
        ],
      ]);
    });

    test('with spaces', async () => {
      const persister = createSqlite3Persister(store, db, {
        serialized: false,
        valuesTable: 'test table',
      });
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'test table',
          'CREATE TABLE "test table"("_id" PRIMARY KEY ON CONFLICT REPLACE)',
          [{_id: '_'}],
        ],
      ]);
    });

    test('with quote', async () => {
      const persister = createSqlite3Persister(store, db, {
        serialized: false,
        valuesTable: 'test "table"',
      });
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'test "table"',
          'CREATE TABLE "test ""table"""("_id" PRIMARY KEY ON CONFLICT REPLACE)',
          [{_id: '_'}],
        ],
      ]);
    });
  });

  describe('Save to empty database', () => {
    test('nothing', async () => {
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("_id" PRIMARY KEY ON CONFLICT REPLACE)',
          [{_id: '_'}],
        ],
      ]);
    });

    test('tables', async () => {
      store.setTables({t1: {r1: {c1: 1}}});
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          't1',
          'CREATE TABLE "t1"("_id" PRIMARY KEY ON CONFLICT REPLACE, "c1")',
          [{_id: 'r1', c1: 1}],
        ],
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("_id" PRIMARY KEY ON CONFLICT REPLACE)',
          [{_id: '_'}],
        ],
      ]);
    });

    test('values', async () => {
      store.setValues({v1: 1});
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("_id" PRIMARY KEY ON CONFLICT REPLACE, "v1")',
          [{_id: '_', v1: 1}],
        ],
      ]);
    });

    test('both', async () => {
      store.setTables({t1: {r1: {c1: 1}}}).setValues({v1: 1});
      await persister.save();
      expect(await getDatabase()).toEqual([
        [
          't1',
          'CREATE TABLE "t1"("_id" PRIMARY KEY ON CONFLICT REPLACE, "c1")',
          [{_id: 'r1', c1: 1}],
        ],
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("_id" PRIMARY KEY ON CONFLICT REPLACE, "v1")',
          [{_id: '_', v1: 1}],
        ],
      ]);
    });
  });

  describe('Load from database', () => {
    test('nothing', async () => {
      await persister.load();
      expect(store.getContent()).toEqual([{}, {}]);
    });

    test('defaulted', async () => {
      await persister.load({t1: {r1: {c1: 1}}}, {v1: 1});
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {v1: 1}]);
    });

    test('broken', async () => {
      setDatabase([
        [
          't1',
          'CREATE TABLE "t1"("di" PRIMARY KEY ON CONFLICT REPLACE, "c1")',
          [{di: 'r1', c1: 1}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{}, {}]);
    });

    test('broken, can default', async () => {
      setDatabase([
        [
          't1',
          'CREATE TABLE "t1"("di" PRIMARY KEY ON CONFLICT REPLACE, "c1")',
          [{di: 'r1', c1: 1}],
        ],
      ]);
      await persister.load({t1: {r1: {c1: 1}}}, {v1: 1});
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {v1: 1}]);
    });

    test('tables', async () => {
      setDatabase([
        [
          't1',
          'CREATE TABLE "t1"("_id" PRIMARY KEY ON CONFLICT REPLACE, "c1")',
          [{_id: 'r1', c1: 1}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {}]);
    });

    test('values', async () => {
      setDatabase([
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("_id" PRIMARY KEY ON CONFLICT REPLACE, "v1")',
          [{_id: '_', v1: 1}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{}, {v1: 1}]);
    });

    test('both', async () => {
      setDatabase([
        [
          't1',
          'CREATE TABLE "t1"("_id" PRIMARY KEY ON CONFLICT REPLACE, "c1")',
          [{_id: 'r1', c1: 1}],
        ],
        [
          'tinybase_values',
          'CREATE TABLE "tinybase_values"("_id" PRIMARY KEY ON CONFLICT REPLACE, "v1")',
          [{_id: '_', v1: 1}],
        ],
      ]);
      await persister.load();
      expect(store.getContent()).toEqual([{t1: {r1: {c1: 1}}}, {v1: 1}]);
    });
  });
});
