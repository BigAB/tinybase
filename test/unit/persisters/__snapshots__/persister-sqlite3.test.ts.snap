// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Save to empty database both 1`] = `
[
  [
    "tinybase",
    [
      {
        "json": "[{"t1":{"r1":{"c1":1}}},{"v1":1}]",
      },
    ],
  ],
]
`;

exports[`Save to empty database nothing 1`] = `
[
  [
    "tinybase",
    [
      {
        "json": "[{},{}]",
      },
    ],
  ],
]
`;

exports[`Save to empty database tables 1`] = `
[
  [
    "tinybase",
    [
      {
        "json": "[{"t1":{"r1":{"c1":1}}},{}]",
      },
    ],
  ],
]
`;

exports[`Save to empty database values 1`] = `
[
  [
    "tinybase",
    [
      {
        "json": "[{},{"v1":1}]",
      },
    ],
  ],
]
`;
